api/pkg/request/auth.go

```diff
@@ -24,11 +26,80 @@ type OIDCVerifier interface {
 	Verify(ctx context.Context, idToken string) (*oidc.IDToken, error)
 }
 
-type MultiOIDCVerifier struct {
+type ClaimsVerifier interface {
+	MatchClaims(ctx context.Context, idToken *oidc.IDToken) error
+}
+
+type OIDCProvider struct {
+	oidcVerifier   OIDCVerifier
+	claimsVerifier ClaimsVerifier
+}
+
+type GithubClaimsVerifier struct {
+	owner  string
+	issuer string
+}
+
+type GithubClaims struct {
+	OIDCSubject string `json:"oidc_sub"`
+	OIDCExtra   string `json:"oidc_extra"`
+	Issuer      string `json:"iss"`
+}
+
+type NilClaimsVerifier struct {
+}
+
+func (d *NilClaimsVerifier) MatchClaims(ctx context.Context, idToken *oidc.IDToken) error {
+	return nil
+}
+
+var DefaultClaimsVerifier = &NilClaimsVerifier{}
+
+func MakeGithubClaimsVerifier(owner string) *GithubClaimsVerifier {
+	return &GithubClaimsVerifier{
+		owner:  owner,
+		issuer: "vstoken.actions.githubusercontent.com",
+	}
+}
+
+func (g *GithubClaimsVerifier) MatchClaims(ctx context.Context, idToken *oidc.IDToken) error {
+	claims := GithubClaims{}
+	err := idToken.Claims(&claims)
+	if err != nil {
+		return errors.Wrap(err, "github id token didn't have expected claims")
+	}
+
+	subject := fmt.Sprintf("repo:%s", g.owner)
+	if !strings.HasPrefix(claims.OIDCSubject, subject) {
+		return errors.Errorf("github id token didn't have the expected oidc_sub, expected to start with %s got %s", subject, claims.OIDCSubject)
+	}
+
+	if claims.Issuer != g.issuer {
+		return errors.Errorf("github id token didn't have the expected issuer, expected %s got %s", claims.Issuer, g.issuer)
+	}
+
+	return nil
+}
+
+func (o *OIDCProvider) Verify(ctx context.Context, rawIDToken string) (*oidc.IDToken, error) {
+	idToken, err := o.oidcVerifier.Verify(ctx, rawIDToken)
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't verify the ID token")
+	}
+
+	err = o.claimsVerifier.MatchClaims(ctx, idToken)
```

Comment:
they have to also pass the configured claims check. If they don't have the right claims it won't work either. This is nice for Github Action flow where the claims contain which owner and repo the Github Action is coming from.